src/oom/oomd-util.c

```diff
@@ -0,0 +1,379 @@
+        return 0;
+}
+
+static int increment_oomd_xattr(const char *path, const char *xattr, uint64_t num_procs_killed) {
+        _cleanup_free_ char *value = NULL;
+        char buf[DECIMAL_STR_MAX(uint64_t) + 1];
+        uint64_t curr_count = 0;
+        int r;
+
+        assert(path);
+        assert(xattr);
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, path, xattr, &value);
+        if (r < 0 && r != -ENODATA)
+                return r;
+
+        if (!isempty(value)) {
+                 r = safe_atou64(value, &curr_count);
+                 if (r < 0)
+                         return r;
+        }
+
+        if (curr_count > UINT64_MAX - num_procs_killed)
+                return -EOVERFLOW;
+
+        xsprintf(buf, "%"PRIu64, curr_count + num_procs_killed);
+        r = cg_set_xattr(SYSTEMD_CGROUP_CONTROLLER, path, xattr, buf, strlen(buf), 0);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+OomdCGroupContext *oomd_cgroup_context_free(OomdCGroupContext *ctx) {
+        if (!ctx)
+                return NULL;
+
+        if (ctx->path)
+                free(ctx->path);
+
+        return mfree(ctx);
+}
+
+void oomd_cgroup_context_hashmap_free(Hashmap *h) {
+        if (!h)
+                return;
+
+        hashmap_free_with_destructor(h, oomd_cgroup_context_free);
+}
+
+const char *oomd_pressure_above(Hashmap *h, unsigned long threshold, usec_t duration, usec_t *ret_last_hit_memory_threshold) {
+        ResourcePressure curr_mem_pressure = {};
+        char *key, *target = NULL;
+        OomdCGroupContext *ctx;
+        usec_t diff;
+        Iterator i;
+
+        assert(ret_last_hit_memory_threshold);
+
+        if (!h) {
+                *ret_last_hit_memory_threshold = 0;
+                return NULL;
+        }
+
+        HASHMAP_FOREACH_KEY(ctx, key, h, i)
+                /* Do a weighted comparison (we care more about 10s, then 60s, then 300s) */
+                if (ctx->memory_pressure.avg10 * 3 + ctx->memory_pressure.avg60 * 2 + ctx->memory_pressure.avg300 >
+                                curr_mem_pressure.avg10 * 3 + curr_mem_pressure.avg60 * 2 + curr_mem_pressure.avg300) {
+                        curr_mem_pressure = ctx->memory_pressure;
+                        target = key;
+                }
+
+        if (curr_mem_pressure.avg10 > threshold) {
+                if (*ret_last_hit_memory_threshold == 0)
+                        *ret_last_hit_memory_threshold = now(CLOCK_MONOTONIC);
+
+                diff = now(CLOCK_MONOTONIC) - *ret_last_hit_memory_threshold;
+                if (diff >= duration)
+                        return target;
+        } else
+                *ret_last_hit_memory_threshold = 0;
+
+        return NULL;
+}
+
+bool oomd_memory_reclaim(Hashmap *h) {
+        uint64_t pgscan = 0, pgscan_of = 0, last_pgscan = 0, last_pgscan_of = 0;
+        OomdCGroupContext *ctx;
+        Iterator i;
+
+        if (!h)
+                return false;
```

Comment:
HASHMAP_FOREACH is happy to be called for NULL hashmaps, it treats them like empty ones, which is what you want here, no?